### ISSUE TITLE
[Hydrated Bloc] Synchronize across isolates (Use bloc data in background isolates)

### ISSUE THREAD

**Issue by zmoshansky** [felangel/bloc#2633]
**Is your feature request related to a problem? Please describe.**
I'm trying to share state between my main isolate and a background callback (isolate).

**Describe the solution you'd like**
Either of the following:
1. The cubit/bloc state should be usable across isolates. (Seems challenging, but I'm new to Flutter/Dart)
2. A manual means of "refreshing" a hydrated bloc to re-load from storage. `reloadFromStorage()`

**Describe alternatives you've considered**
1. Using shared preferences
2. Trying to use Hive directly
3. Using Hydrated Bloc: Restarting app, causing the initial load of data to pull the new values. Needed on every switch between updating a value via UI, and a background callback having the updated value.

**Additional context**
In both main and a background isolate callback, I ensure that I initialize storage as follows. Each isolate then creates an instance of my `SettingsCubit` which contains state I'd like to share between my UI and a background isolate (long running timers)

```
initStorage() async {
  try {
    HydratedBloc.storage;
    debugPrint(
        '[${DateTime.now()}] Isolate: ${Isolate.current.hashCode} bloc storage: initialized already');
  } catch (error) {
    debugPrint(
        '[${DateTime.now()}] Isolate: ${Isolate.current.hashCode} bloc storage: initializing');
    HydratedBloc.storage = await HydratedStorage.build(
      storageDirectory: kIsWeb
          ? HydratedStorage.webStorageDirectory
          : await getTemporaryDirectory(),
    );
  }
}
```

**Comment by felangel** [comment]
Hi @zmoshansky 👋 
Thanks for opening an issue!

Can you please provide a link to a minimal reproduction sample that illustrates the problem you're facing? I'm happy to take a closer look and provide feedback/suggestions. Thanks 🙏

**Comment by zmoshansky** [comment]
It'll probably take a while to get a repo up, but here's the gist of the code... 

Main:
```
late final backgroundScheduler = BackgroundScheduler();

Future<void> main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await initStorage();

  backgroundScheduler.initialize();
  runApp(MyApp());
}
```

Background Scheduler:
```
alarmCallback() async {
  await initStorage();
  final settings = settingsSingleton;
  print('settings=${settings.toJson(settings.state)}');
}

class BackgroundScheduler {
  final settings = settingsSingleton;
  late final StreamSubscription<SettingsState> settingsSub;
  var initializedPlugin = false;

  initialize() {
    setAlarms(settings.state);
    settingsSub = settings.stream.listen(setAlarms);
  }

  setAlarms(SettingsState settingsState) async {
    if (!initializedPlugin) {
      initializedPlugin = await AndroidAlarmManager.initialize();
    }
   
    final int alarmID = 0;
      await AndroidAlarmManager.periodic(
          Duration(minutes: settingsState.interval), alarmID, alarmCallback);

      // Change a value; Would expect this to reflect in background use of `settingsSingleton`
      settings.setEnabled(true);

  }

  dispose() async {
    await settingsSub.cancel();
  }
}
```

Cubit
```
import 'package:hydrated_bloc/hydrated_bloc.dart';

final SettingsCubit settingsSingleton = SettingsCubit();

class SettingsCubit extends HydratedCubit<SettingsInitial> {
  SettingsCubit() : super(SettingsInitial(false));

  setEnabled(bool enabled) => emit(state.copyWith(enabled: enabled));
  @override
  SettingsInitial? fromJson(Map<String, dynamic> json) => SettingsInitial(
        json['value']['enabled'] as bool,
      );

  @override
  Map<String, dynamic>? toJson(SettingsState state) => {
        'value': {
          'enabled': state.enabled,
        }
      };
}

abstract class SettingsState {
  late final bool enabled;
}

class SettingsInitial extends SettingsState {
  final enabled;

  SettingsInitial(this.enabled);

  SettingsInitial copyWith({
    bool? enabled,
  }) {
    return SettingsInitial(
      enabled ?? this.enabled,
    );
  }
}
```

Widget Tree:
```
return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      drawer: AppDrawer(),
      body: Center(
        child: BlocProvider(
          lazy: false,
          create: (BuildContext context) => settingsSingleton,
          child: ListView(children: <Widget>[
            //....
          ]),
        ),
      ),
    );
    ```

**Comment by squallsama** [comment]
Have you tried https://github.com/Maksimka101/isolate-bloc/tree/master/packages/isolate_bloc ?

**Comment by heshaShawky** [comment]
I hope if there's a way to run the bloc logic in isolate instead of recreating the whole thing again in isolated function

**Comment by federicodesia** [comment]
Hi, I have the same problem. Is there any update on this?

**Comment by felangel** [comment]
It would be very helpful if anyone could provide a minimal reproduction sample, thanks!

**Comment by federicodesia** [comment]
Suppose an app uses Firebase Cloud Messaging to notify the background app when it needs to update the state of the HydratedCubit. Background message handling is done in a new isolate, so we cannot access the current HydratedCubit instance that was created in the `BlocProvider` within the root thread.

What we can do is create a new instance of HydratedCubit inside the new isolate using the same storage directory, which will emit the required state, then run the `toJson` function, saving it to Hive storage. But when the application is resumed from the background, the state was not updated in the root instance.

This is where a function like `reloadFromStorage()` would allow updating the state by taking the stored Json, this is very similar to the reverse of the `hydrate()` function that can be used with `HydratedMixin`.

I add an example repository which you can find [here](https://github.com/federicodesia/hydrated_isolate_example). To simulate the notification messages I have created a fake BackgroundWorker, which is executed when the application is minimized emitting a new state.

Steps to reproduce:
1. Press the button to count up or down.
2. Minimize the application (a new instance of the HydratedCubit will emit a state that sets the counter to zero from an isolate)
3. Maximize the application (counter state is expected to be zero)

I want to thank you for the disposition and all the work that you and the community do for this package.

**Comment by zmoshansky** [comment]
@federicodesia Thank you very much for adding a minimal repro!
@felangel I can second that a `reloadFromStorage` function would allow us to work around a lot of cross-isolate issues by using hive (HydratedStorage) as an intermediary in cases where the different isolates access data in a mutually exclusive manner.

My thoughts would be to modify `hydrated_bloc.dart`, somewhat like the following...
https://github.com/felangel/bloc/blob/dd65f69997087dd3250403db286082d04f983b3e/packages/hydrated_bloc/lib/src/hydrated_bloc.dart#L216

```
Function<void> reloadFromStorage() {
    try {
      final stateJson = _storage.read(storageToken) as Map<dynamic, dynamic>?;
      if (stateJson == null) {
        _state = super.state;
      }
      final cachedState = _fromJson(stateJson);
      if (cachedState == null) {
        _state = super.state;
      }
      _state = cachedState;
     super.emit(_state)
    } catch (error, stackTrace) {
      onError(error, stackTrace);
      _state = super.state;
     super.emit(_state)
    }
```

Automating this across isolates using the same method wouldn't be too difficult (Extend resulting class and use a Port/Channel to communicate when any isolate changes the data). The obvious caveat that sending data through the DB is not the most performant way, but for mutually exclusive access, it's a start.

**Comment by federicodesia** [comment]
I tried something similar to what @zmoshansky mentions but with no success. I found that using Hive with multiple readers or writers at the same time brings many problems, and when it comes to different Isolates there are even more communication problems that are external to HydratedBloc.

I am currently using an implementation I made with the Drift database that is a bit more prepared than Hive for the use of multiple Isolates and allows to listen to changes from any of them. You can find it [here](https://github.com/federicodesia/task_manager/tree/main/lib/blocs/drifted_bloc), though I haven't done much work on it yet and it should be improved for real use as it's not really optimized.

Also, I think this issue should be kept open as it is a very typical use case in many applications and should be considered in the package.

**Comment by Rexios80** [comment]
Tracking isolate support for Hive CE here: https://github.com/IO-Design-Team/hive_ce/issues/69

**Comment by Rexios80** [comment]
Good news! I have made great progress on isolate support for Hive CE. This will require a new implementation of `HydratedStorage` that uses `IsolatedHive` since isolate communication is asynchronous. Maybe `Hive` could be replaced with `IsolatedHive` in the default `HydratedStorage`? We might have to do benchmarks to ensure that makes sense since isolate communication introduces overhead. The PR for isolate support (https://github.com/IO-Design-Team/hive_ce/pull/80) still needs a lot of work to be mergeable, but I wanted to update you all on the progress!

**Comment by Rexios80** [comment]
Isolate support for Hive CE has been pre-released! The only reason this isn't a real release is because I can't test it anymore myself, and I want to make sure there are no major issues first. I have made a sample of what `IsolatedHydratedStorage` might look like below. The only caveat is you _must_ pass an `isolateNameServer` for this to actually be safe in a multi-isolate environment. I have also pasted the `IsolateNameServer` from `hive_ce_flutter` as an example. I'm not sure how `hydrated_bloc` would handle this. Maybe there should be a `hydrated_bloc_flutter`?

<details>
<summary>IsolatedHydratedStorage</summary>

```dart
import 'dart:convert';
import 'dart:io';

import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:hive_ce/hive.dart';
import 'package:hydrated_bloc/hydrated_bloc.dart';
import 'package:path/path.dart' as path;

// Prevents conflicts with multiple Hive inits
// ignore: implementation_imports
import 'package:hive_ce/src/isolate/isolated_hive_impl/isolated_hive_impl.dart';

class IsolatedHydratedStorage implements Storage {
  @visibleForTesting
  IsolatedHydratedStorage(this._box);

  static Future<IsolatedHydratedStorage> build({
    required HydratedStorageDirectory storageDirectory,
    HydratedCipher? encryptionCipher,
    IsolateNameServer? isolateNameServer,
  }) {
    return _lock.synchronized(() async {
      hive = IsolatedHiveImpl();
      IsolatedBox<dynamic> box;

      if (storageDirectory == HydratedStorageDirectory.web) {
        box = await hive.openBox<dynamic>(
          'hydrated_box',
          encryptionCipher: encryptionCipher,
        );
      } else {
        await hive.init(
          storageDirectory.path,
          isolateNameServer: isolateNameServer,
        );
        box = await hive.openBox<dynamic>(
          'hydrated_box',
          encryptionCipher: encryptionCipher,
        );
        await migrate(storageDirectory.path, box);
      }

      return IsolatedHydratedStorage(box);
    });
  }

  @visibleForTesting
  static late IsolatedHiveInterface hive;

  final IsolatedBox<dynamic> _box;

  @override
  dynamic read(String key) => _box.isOpen ? _box.get(key) : null;

  @override
  Future<void> write(String key, dynamic value) async {
    if (!_box.isOpen) return;
    return _box.put(key, value);
  }

  @override
  Future<void> delete(String key) async {
    if (!_box.isOpen) return;
    return _box.delete(key);
  }

  @override
  Future<void> clear() async {
    if (!_box.isOpen) return;
    await _box.clear();
  }

  @override
  Future<void> close() async {
    if (!_box.isOpen) return;
    return _box.close();
  }
}

Future<dynamic> migrate(String directory, IsolatedBox<dynamic> box) async {
  final file = File(path.join(directory, '.hydrated_bloc.json'));
  if (file.existsSync()) {
    try {
      final dynamic storageJson = json.decode(await file.readAsString());
      final cache = (storageJson as Map).cast<String, String>();
      for (final key in cache.keys) {
        try {
          final string = cache[key];
          final dynamic object = json.decode(string ?? '');
          await box.put(key, object);
        } catch (_) {}
      }
    } catch (_) {}
    await file.delete();
  }
}

```
</details>

<details>
<summary>IsolateNameServer</summary>

```dart
import 'dart:isolate';
import 'dart:ui' as flutter;

import 'package:hive_ce/hive.dart' as hive;

/// A wrapper around [flutter.IsolateNameServer] for [IsolatedHive]
class IsolateNameServer extends hive.IsolateNameServer {
  /// Constructor
  const IsolateNameServer();

  @override
  SendPort? lookupPortByName(String name) =>
      flutter.IsolateNameServer.lookupPortByName(name);

  @override
  bool registerPortWithName(SendPort port, String name) =>
      flutter.IsolateNameServer.registerPortWithName(port, name);

  @override
  bool removePortNameMapping(String name) =>
      flutter.IsolateNameServer.removePortNameMapping(name);
}

```
</details>

**Comment by Rexios80** [comment]
@felangel I am planning on making a stable release of `IsolatedHive` soon. Do you have any comments or questions?

**Comment by felangel** [comment]
> [@felangel](https://github.com/felangel) I am planning on making a stable release of `IsolatedHive` soon. Do you have any comments or questions?

I haven’t had a chance to try it but I’ll try to make some time later today 👍

**Comment by Rexios80** [comment]
Isolate support is now stable in HIve CE! I do not think there is a way around creating a hydrated_bloc_flutter package if we want automatic setup of IsolatedHive. We might also want a version of IsolatedHydratedStorage in normal hydrated_bloc that users can pass their own IsolateNameServer into in case they have a use-case for it.

**Comment by TijnvandenEijnde** [comment]
@felangel Thank you for the package! Any updates on this?

**Comment by felangel** [comment]
@TijnvandenEijnde I believe this should be unblocked because you should already be able to create your own `IsolatedHydratedStorage` by implementing the `Storage` interface and using the `IsolatedHive` support from `hive_ce`. Let me know if you run into issues, thanks!

**Comment by TijnvandenEijnde** [comment]
@felangel Thanks for the quick response! I already started working on a fork to possibly merge into Hydrated Bloc: https://github.com/TijnvandenEijnde/bloc/tree/feat/hydrated-bloc-synchronize-across-isolates. However, I will give your suggestion a try. I will let you know if I run into any issues, thanks!

**Comment by TijnvandenEijnde** [comment]
@felangel Your suggested approach is working, but I think there is one problem.

```dart
import 'dart:async';

import 'package:hive_ce_flutter/adapters.dart';
import 'package:hydrated_bloc/hydrated_bloc.dart';
import 'package:synchronized/synchronized.dart';

class HydratedIsolatedStorage implements Storage {
  HydratedIsolatedStorage(this._box);

  static Future<HydratedIsolatedStorage> build({
    required HydratedStorageDirectory storageDirectory,
    HydratedCipher? encryptionCipher,
  }) {
    return _lock.synchronized(() async {
      await IsolatedHive.initFlutter(subDirectory: storageDirectory.path);
      final box = await IsolatedHive.openBox<dynamic>(
        'hydrated_box',
        encryptionCipher: encryptionCipher,
      );

      return HydratedIsolatedStorage(box);
    });
  }

  static final _lock = Lock();

  final IsolatedBox<dynamic> _box;

  @override
  Future<dynamic> read(String key) async =>
      _box.isOpen ? await _box.get(key) : null;

  @override
  Future<void> write(String key, dynamic value) async {
    if (_box.isOpen) return _lock.synchronized(() => _box.put(key, value));
  }

  @override
  Future<void> delete(String key) async {
    if (_box.isOpen) return _lock.synchronized(() => _box.delete(key));
  }

  @override
  Future<void> clear() async {
    if (_box.isOpen) return _lock.synchronized(_box.clear);
  }

  @override
  Future<void> close() async {
    if (_box.isOpen) return _lock.synchronized(_box.close);
  }
}

```

As you can see I have to make the `read` override a Future because the [get](https://github.com/IO-Design-Team/hive_ce/blob/c36ec41aa160801676e944668aa26a79a53f1f6f/hive/lib/src/isolate/isolated_box.dart#L78) function of the `IsolatedBox` class is a future. I don't think we can ignore this.

This will mean that the `Storage` class needs to be adjusted and therefore the `hydrate` function of the `HydratedBloc` class will become asynchronous as well. 

I already tried implementing it in this [branch](https://github.com/TijnvandenEijnde/bloc/tree/feat/add-support-for-isolated-hive). However, now some of the `e2e_test` tests are failing and throw: 

```
⚠️ WARNING: HIVE MULTI-ISOLATE RISK DETECTED ⚠️

Accessing Hive from an unsafe isolate (current isolate: "test_suite:file:///var/folders/qm/c57wvs656n956fywgyygm99m0000gq/T/dart_test.kernel.bV4o0J/test.dart_1.dill")
This can lead to DATA CORRUPTION as Hive boxes are not designed for concurrent
access across isolates. Each isolate would maintain its own box cache,
potentially causing data inconsistency and corruption.

RECOMMENDED ACTIONS:
- Use IsolatedHive instead
```  

I was already able to make the tests succeed, but I don't think that I can get rid of the warning.

So I am a bit confused what the right approach is here.  

Also, @Rexios80 do you think the approach I took is similar to what you had in mind?

**Comment by Rexios80** [comment]
Related:
- https://github.com/IO-Design-Team/hive_ce/issues/242

**Comment by Rexios80** [comment]
@TijnvandenEijnde You can now mute this by calling `HiveLogger.unsafeIsolateWarning = false`

**Comment by TijnvandenEijnde** [comment]
@Rexios80 Great, thank you for letting me know!
